File: p42626.java
import java.util.PriorityQueue;

public class p42626 {
    public static int solution(int[] scoville, int K) {
        int answer = 0;
        PriorityQueue<Integer> pq = new PriorityQueue<>();
        for(int s : scoville) pq.add(s);

        while(pq.peek() < K && pq.size() >= 2){
            int min1 = pq.poll();
            int min2 = pq.poll();
            int new_scv = min1 + (min2 * 2);
            answer++;
            pq.add(new_scv);
        }
        if(pq.isEmpty() || pq.peek() < K) answer = -1;
        return answer;
    }

    public static void main(String[] args){
        System.out.println(solution(new int[]{1, 2, 3, 9, 10, 12}, 7));
    }
}

File: p42626.cpp
#include<bits/stdc++.h>
using namespace std;

int solution(vector<int> scoville, int K) {
    int answer = 0;
    priority_queue<int, vector<int>, greater<int>> pq(scoville.begin(), scoville.end());
    
    int n = 0;

    while(pq.size() >= 2 && pq.top() < K){
        n = pq.top();
        pq.pop();
        n += pq.top() * 2;
        pq.pop();
        pq.push(n);
        answer++;
    }

    return n >= K ? answer : -1;
}